nomad/client_rpc.go

```diff
@@ -0,0 +1,123 @@
+package nomad
+
+import (
+	"fmt"
+	"time"
+
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/yamux"
+)
+
+// nodeConnState is used to track connection information about a Nomad Client.
+type nodeConnState struct {
+	// Session holds the multiplexed yamux Session for dialing back.
+	Session *yamux.Session
+
+	// Established is when the connection was established.
+	Established time.Time
+}
+
+// getNodeConn returns the connection to the given node and whether it exists.
+func (s *Server) getNodeConn(nodeID string) (*nodeConnState, bool) {
+	s.nodeConnsLock.RLock()
+	defer s.nodeConnsLock.RUnlock()
+	state, ok := s.nodeConns[nodeID]
+	return state, ok
+}
+
+// connectedNodes returns the set of nodes we have a connection with.
+func (s *Server) connectedNodes() map[string]time.Time {
+	s.nodeConnsLock.RLock()
+	defer s.nodeConnsLock.RUnlock()
+	nodes := make(map[string]time.Time, len(s.nodeConns))
+	for nodeID, state := range s.nodeConns {
+		nodes[nodeID] = state.Established
+	}
+	return nodes
+}
+
+// addNodeConn adds the mapping between a node and its session.
+func (s *Server) addNodeConn(ctx *RPCContext) {
+	// Hotpath the no-op
+	if ctx == nil || ctx.NodeID == "" {
+		return
+	}
+
+	s.nodeConnsLock.Lock()
+	defer s.nodeConnsLock.Unlock()
+	s.nodeConns[ctx.NodeID] = &nodeConnState{
+		Session:     ctx.Session,
+		Established: time.Now(),
+	}
+}
+
+// removeNodeConn removes the mapping between a node and its session.
+func (s *Server) removeNodeConn(ctx *RPCContext) {
+	// Hotpath the no-op
+	if ctx == nil || ctx.NodeID == "" {
+		return
+	}
+
+	s.nodeConnsLock.Lock()
+	defer s.nodeConnsLock.Unlock()
+	delete(s.nodeConns, ctx.NodeID)
+}
+
+// serverWithNodeConn is used to determine which remote server has the most
+// recent connection to the given node. The local server is not queried.
+// ErrNoNodeConn is returned if all local peers could be queried but did not
+// have a connection to the node. Otherwise if a connection could not be found
+// and there were RPC errors, an error is returned.
+func (s *Server) serverWithNodeConn(nodeID string) (*serverParts, error) {
+	s.peerLock.RLock()
+	defer s.peerLock.RUnlock()
+
+	// We skip ourselves.
+	selfAddr := s.LocalMember().Addr.String()
```

Comment:
Would it be better to use the ID of the server to skip itself rather than the address? That's the direction we are headed in with Raft 3 anyway so I think we should not add more conditional code that uses the address of the server to make decisions rather than the ID which is immutable.

Reply:
The local peers is mapped by address so this is the simplest approach.